core/src/main/scala/kafka/controller/ReplicaStateMachine.scala

```diff
@@ -192,9 +192,11 @@ class ZkReplicaStateMachine(config: KafkaConfig,
 
           currentState match {
             case NewReplica =>
-              val assignment = controllerContext.partitionReplicaAssignment(partition)
-              if (!assignment.contains(replicaId)) {
-                controllerContext.updatePartitionReplicaAssignment(partition, assignment :+ replicaId)
+              val assignment = controllerContext.partitionFullReplicaAssignment(partition)
+              if (!assignment.replicas.contains(replicaId)) {
+                error(s"Adding replica ($replicaId) that is not part of the assignment $assignment")
+                val newAssignment = assignment.copy(replicas = assignment.replicas :+ replicaId)
```

Comment:
This code smells a bit (not from this PR). Feels like we need to refetch the assignment state in order to confirm whether the replicaId should be there or not. I guess after zk removal, the controller context will be guaranteed to be up to date and we can improve this.

Reply:
Yeah. This state machine is a little strange. I convinced myself that this condition is not possible. The controllers needs to find out that `replicaId` is assigned to `partition` to be able to call `doHandleStateChanges` with those values and this is only possible if `controllerContext` has this information.

Another insight to why this code path is not possible is that this condition would magically change the ordering of the assignment.

I made the conservative change and simply logged an error message. If we are interested, I can revisit this code to remove this code path.

Reply:
Yeah, we don't need to do it here, but if you think the state is not reachable, I'd suggest filing a separate PR. I'd rather raise an error instead of trying to fix up some state inconsistency. It's a bit arbitrary to assume the state change is probably right.